### DISCUSSION TITLE
Show key type info when hovering over an index lookup on TypedDicts

### DISCUSSION THREAD

**Discussion by R-O-C-K-E-T** [microsoft/pylance-release#4720] [Enhancement]
## Current Behaviour
When hovering over an index operation into a `TypedDict` variable no type info for the key appears and definition is unavailable.

In contrast constructing a `TypedDict` object hovering over keys there does produce type info and definition.

```
from typing import TypedDict

class Test(TypedDict):
    field: str

test: Test = {
    'field': 'value',  # Hovering over 'field' gives type info and the definition is available
}
test['field']  # Hovering over 'field' in this case gives no info and no definition is found
```

## Expected Behaviour
Hovering over a `TypedDict` key within an indexing operation should show the key type info and the definition should be available.
![Screenshot from 2023-08-06 14-12-01](https://github.com/microsoft/pyright/assets/2784602/aa4ae276-d503-4f29-8008-6b5b1238cfcf)

## Motivating Use Case
Plain dictionaries often appear in legacy code bases and when migrating types into these code bases it is easier and less error prone to write `TypedDict`s to describe these dictionaries instead of migrating them into dataclasses or similar (IMO). Which do show type info when hovering over the property names.

Migrating this way works well to verify that all accesses and definitions are valid. However when writing code that indexes into one of these dictionaries, finding either the type, docstring or definition of a key is cumbersome. Since the developer must find the type of the indexed object the and manually scan through to the relevant key.

### Hover/Definition search for `TypedDict.get`
`.get` is often used when keys are marked `NotRequired` to skip verbose `'key' in obj` checks. It may improve usability if key definitions were easily available.

```python
from typing import TypedDict, NotRequired

class Test(TypedDict):
    field: NotRequired[str]

test: Test = {}
test.get('field')  # Hovering over 'field' could show type info for the 'field' key
```

### Hover/Definition search for `NamedTuple` indexing
The same functionality for `TypedDict` could be applied to `NamedTuple`. The use case is weaker however, since the field can be directly referenced by name instead of by index.

```python
from typing import NamedTuple

class Test(NamedTuple):
	first: int
	second: str

test = Test(1, 'second')
res = test[1]  # Hovering over 1 could show type info for the 'second' key
```

**Comment by erictraut** [comment]
The hover provider shows the types of identifiers in your code. It doesn't do anything for a string literal because the type of that string literal is, well, a `str`. I suppose it could treat string literals specially if they are used to index a TypedDict, but that would be a special case and somewhat inconsistent.

In any case, I will transfer this request to the pylance-release project since it relates to a language server feature. The pylance team can decide whether to pursue this enhancement request or wait for additional feedback from pylance users.

@rchiodo or @debonte, could you transfer this to pylance-release? Thanks!

**Comment by debonte** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
